src/instance.rs

```diff
@@ -214,6 +225,123 @@ unsafe impl<T> AsPyPointer for Py2<'_, T> {
     }
 }
 
+/// A borrowed equivalent to `Py2`.
+///
+/// The advantage of this over `&Py2` is that it avoids the need to have a pointer-to-pointer, as Py2
+/// is already a pointer to a PyObject.
+#[repr(transparent)]
+pub(crate) struct Py2Borrowed<'a, 'py, T>(NonNull<ffi::PyObject>, PhantomData<&'a Py2<'py, T>>); // TODO is it useful to have the generic form?
+
+impl<'a, 'py> Py2Borrowed<'a, 'py, PyAny> {
```

Comment:
Should at least this block or even the struct definition itself have a bound ensuring that `'py` outlives `'a` to avoid all to obvious lifetime errors?

Reply:
I think it's an edge case; isn't this a valid construction where `'a` outlives `'py`:

```rust
let any: Py<PyAny> = /* ... */;
let any = &any;  // start borrow 'a
Python::with_gil(|py| Py2Borrowed::from(any.attach(py)));  // 'py shorter than 'a
```

Maybe in practice this is never useful though?

Reply:
I think in this case, the borrow of `any` could always be shortened to at most `'py` by reborrowing. So if the bound does not prevent the above but prevents some illegal uses, I would prefer to have it.

Reply:
Actually it turns out that exactly in the `Py<PyBytes>::as_bytes` case we're looking at in #3651 it's the case that we want `'a` to outlive `'py`: https://github.com/PyO3/pyo3/pull/3651/commits/d6f18b5612ab2873e3abd5d91487d79019918497

Reply:
(I am also sort of confused now whether we discuss #3651 or #3654. Maybe these should be folded into #3651 with `Py2Borrowed` just being a separate commit?)